<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="pendulum">
  <xacro:macro name="flywheel" params="radius length offset mass motor_mass">
    <link name="flywheel_link">
      <visual>
        <geometry>
          <cylinder length="${length}" radius="${radius}"/>
        </geometry>
      </visual>
      <collision>
        <geometry>
          <cylinder length="${length}" radius="${radius}"/>
        </geometry>
      </collision>
      <inertial>
        <mass value="${mass + motor_mass}"/>
        <inertia ixx="${1/12*mass*(3*radius**2 + length**2)}"
                 ixy="0.0"
                 ixz="0.0"
                 iyy="${1/12*mass*(3*radius**2 + length**2)}"
                 iyz="0.0"
                 izz="${1/2*mass*(radius**2)}"/>
      </inertial>
    </link>
    
    <joint name="arm_to_flywheel" type="continuous">
      <axis xyz="0 0 1"/>
      <parent link="arm_link"/>
      <child link="flywheel_link"/>
      <origin rpy="0 ${3.1415/2} 0" xyz="0 0 ${offset}"/>
    </joint>
    
    <transmission name="$flywheel_transmission">
      <type>transmission_interface/SimpleTransmission</type>
      <actuator name="$flywheel_motor">
        <mechanicalReduction>1</mechanicalReduction>
      </actuator>
      <joint name="arm_to_flywheel">
        <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
      </joint>
    </transmission>
  </xacro:macro>
</robot>

